pkg/operator/api/api.go

```diff
@@ -63,72 +62,81 @@ type apiServer struct {
 	// notify all dapr runtime
 	connLock          sync.Mutex
 	allConnUpdateChan map[string]chan *componentsapi.Component
+	ready             chan struct{}
 }
 
 // NewAPIServer returns a new API server.
 func NewAPIServer(client client.Client) Server {
 	return &apiServer{
 		Client:            client,
 		allConnUpdateChan: make(map[string]chan *componentsapi.Component),
+		ready:             make(chan struct{}),
 	}
 }
 
 // Run starts a new gRPC server.
-func (a *apiServer) Run(ctx context.Context, certChain *daprCredentials.CertChain, onReady func()) {
+func (a *apiServer) Run(ctx context.Context, certChain *daprCredentials.CertChain) error {
+	select {
+	case <-a.ready:
+		return errors.New("api server already running")
+	default:
+	}
+
 	log.Infof("starting gRPC server on port %d", serverPort)
 
 	opts, err := daprCredentials.GetServerOptions(certChain)
 	if err != nil {
-		log.Fatalf("error creating gRPC options: %v", err)
+		return fmt.Errorf("error creating gRPC options: %v", err)
 	}
 	s := grpc.NewServer(opts...)
 	operatorv1pb.RegisterOperatorServer(s, a)
 
 	lis, err := net.Listen("tcp", fmt.Sprintf(":%v", serverPort))
 	if err != nil {
-		log.Fatalf("error starting tcp listener: %v", err)
+		return fmt.Errorf("error starting tcp listener: %v", err)
 	}
 
-	if onReady != nil {
-		onReady()
-	}
+	close(a.ready)
```

Comment:
Still unclear why the channel is closed if the goal is to use it as semaphore

Reply:
A closed ready channel signals that the apiServer is ready.

Reply:
But then you don’t have a semaphore anymore to prevent others from calling the method… or not?

Reply:
A new `apiServer` will have a `ready` channel which is not closed. This means the code block 

```go
	select {
	case <-a.ready:
		return errors.New("api server already running")
	default:
	}
```

will reach the default case and not error. When a listener has been created successfully, the ready channel will be closed. The above select statement will then entry the channel case and error.